internal/router/middlewares/otelhttp.go

```diff
@@ -3,12 +3,23 @@ package middlewares
 import (
 	"net/http"
 
+	"github.com/textileio/go-tableland/pkg/metrics"
 	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
 )
 
 // OtelHTTP wraps the handler h with OTEL metrics.
 func OtelHTTP(operation string) func(h http.Handler) http.Handler {
 	return func(h http.Handler) http.Handler {
-		return otelhttp.NewHandler(h, operation)
+		return otelhttp.NewHandler(&labeledHandler{h: h}, operation)
 	}
 }
+
+type labeledHandler struct {
+	h http.Handler
+}
+
+func (lh *labeledHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	labeler, _ := otelhttp.LabelerFromContext(r.Context())
```

Comment:
The `otelhttp.NewHandler` in L13 creates a `http.Handler` that internally creates an internal `&Labeler{}`.

That means that in wrapped internal `http.Handlers`, middlewares can do `otelhttp.LabelerFromContext(...)` can get that "base labeler". That allows each handler to add custom extra attributes to be included in captured metrics.

This is exactly what we need. What we do is always include the `service_name` form the new `metrics.BaseAttrs`.

The same mechanics might be useful if we want to include custom labels in different handlers.  For example, we could try including the `chain_id`, so we can have all the `otelhttp` metrics we have today, but also separated by chain_id, which we don't have.

We can think if that's useful in the future, it would be an easy addition. I'd have to just do both L22-L23-ish in the "chain id" detection middleware.

Reply:
Nice!